src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleSerializer.cs

```diff
@@ -0,0 +1,155 @@
+using System.Linq;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Hl7.Fhir.Utility;
+using Microsoft.Health.Fhir.Core.Extensions;
+using Microsoft.Health.Fhir.Shared.Core.Features.Search;
+
+namespace Microsoft.Health.Fhir.Api.Features.Resources.Bundle
+{
+    public class BundleSerializer
+    {
+        private readonly JsonWriterOptions _writerOptions = new JsonWriterOptions { Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping };
+
+        public BundleSerializer()
+        {
+        }
+
+        public async Task Serialize(Hl7.Fhir.Model.Bundle bundle, Stream outputStream)
+        {
+            using (Utf8JsonWriter writer = new Utf8JsonWriter(outputStream, _writerOptions))
+            using (StreamWriter streamWriter = new StreamWriter(outputStream, leaveOpen: true))
+            {
+                writer.WriteStartObject();
+
+                writer.WriteString("resourceType", bundle.ResourceType.GetLiteral());
+                writer.WriteString("id", bundle.Id);
+
+                SerializeMetadata();
+
+                writer.WriteString("type", bundle.Type?.GetLiteral());
+
+                SerializeLinks();
+
+                if (bundle.Total.HasValue)
+                {
+                    writer.WriteNumber("total", bundle.Total.Value);
+                }
+
+                await SerializeEntries();
+
+                writer.WriteEndObject();
+                await writer.FlushAsync();
+
+                void SerializeMetadata()
+                {
+                    if (bundle.Meta != null)
+                    {
+                        writer.WriteStartObject("meta");
+                        writer.WriteString("lastUpdated", bundle.Meta?.LastUpdated?.ToInstantString());
+                        writer.WriteEndObject();
+                    }
+                }
+
+                void SerializeLinks()
+                {
+                    if (bundle.Link?.Any() == true)
+                    {
+                        writer.WriteStartArray("link");
+
+                        foreach (var link in bundle.Link)
+                        {
+                            writer.WriteStartObject();
+                            writer.WritePropertyName("relation");
+                            writer.WriteStringValue(link.Relation);
+                            writer.WritePropertyName("url");
+                            writer.WriteStringValue(link.Url);
+                            writer.WriteEndObject();
+                        }
+
+                        writer.WriteEndArray();
+                    }
+                }
+
+                async Task SerializeEntries()
+                {
+                    if (bundle.Entry?.Any() == true)
+                    {
+                        writer.WriteStartArray("entry");
+                        foreach (var entry in bundle.Entry)
+                        {
+                            if (!(entry is RawBundleEntryComponent))
```

Comment:
You can combine this and the cast into one statement 
``` csharp
if (!(entry is RawBundleEntryComponent rawBundleEntry))
```

Reply:
Thanks! Learned something new :)